table/snapshot_producers.go

```diff
@@ -0,0 +1,582 @@
+	return fmt.Sprintf("%s-m%d.avro", commit, num)
+}
+
+func newManifestListFileName(snapshotID int64, attempt int, commit uuid.UUID) string {
+	// mimics behavior of java
+	// https://github.com/apache/iceberg/blob/c862b9177af8e2d83122220764a056f3b96fd00c/core/src/main/java/org/apache/iceberg/SnapshotProducer.java#L491
+	return fmt.Sprintf("snap-%d-%d-%s.avro", snapshotID, attempt, commit)
+}
+
+func newFastAppendFilesProducer(op Operation, txn *Transaction, fs iceio.WriteFileIO, commitUUID *uuid.UUID, snapshotProps iceberg.Properties) *snapshotProducer {
+	prod := createSnapshotProducer(op, txn, fs, commitUUID, snapshotProps)
+	prod.producerImpl = &fastAppendFiles{base: prod}
+
+	return prod
+}
+
+type fastAppendFiles struct {
+	base *snapshotProducer
+}
+
+func (fa *fastAppendFiles) processManifests(manifests []iceberg.ManifestFile) ([]iceberg.ManifestFile, error) {
+	return manifests, nil
+}
+
+func (fa *fastAppendFiles) existingManifests() ([]iceberg.ManifestFile, error) {
+	existing := make([]iceberg.ManifestFile, 0)
+	if fa.base.parentSnapshotID > 0 {
+		previous, err := fa.base.txn.meta.SnapshotByID(fa.base.parentSnapshotID)
+		if err != nil {
+			return nil, fmt.Errorf("could not find parent snapshot %d", fa.base.parentSnapshotID)
+		}
+
+		manifests, err := previous.Manifests(fa.base.io)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, m := range manifests {
+			if m.HasAddedFiles() || m.HasExistingFiles() || m.SnapshotID() == fa.base.snapshotID {
+				existing = append(existing, m)
+			}
+		}
+	}
+
+	return existing, nil
+}
+
+func (fa *fastAppendFiles) deletedEntries() ([]iceberg.ManifestEntry, error) {
+	return nil, nil
+}
+
+type mergeAppendFiles struct {
+	fastAppendFiles
+
+	targetSizeBytes int
+	minCountToMerge int
+	mergeEnabled    bool
+}
+
+func newMergeAppendFilesProducer(op Operation, txn *Transaction, fs iceio.WriteFileIO, commitUUID *uuid.UUID, snapshotProps iceberg.Properties) *snapshotProducer {
+	prod := createSnapshotProducer(op, txn, fs, commitUUID, snapshotProps)
+	prod.producerImpl = &mergeAppendFiles{
+		fastAppendFiles: fastAppendFiles{base: prod},
+		targetSizeBytes: snapshotProps.GetInt(ManifestTargetSizeBytesKey, ManifestTargetSizeBytesDefault),
+		minCountToMerge: snapshotProps.GetInt(ManifestMinMergeCountKey, ManifestMinMergeCountDefault),
+		mergeEnabled:    snapshotProps.GetBool(ManifestMergeEnabledKey, ManifestMergeEnabledDefault),
+	}
+
+	return prod
+}
+
+func (m *mergeAppendFiles) processManifests(manifests []iceberg.ManifestFile) ([]iceberg.ManifestFile, error) {
+	unmergedDataManifests, unmergedDeleteManifests := []iceberg.ManifestFile{}, []iceberg.ManifestFile{}
+	for _, m := range manifests {
+		if m.ManifestContent() == iceberg.ManifestContentData {
+			unmergedDataManifests = append(unmergedDataManifests, m)
+		} else if m.ManifestContent() == iceberg.ManifestContentDeletes {
+			unmergedDeleteManifests = append(unmergedDeleteManifests, m)
+		}
```

Comment:
What's the behavior here when it doesn't match? should we throw?

Reply:
I was following the pyiceberg example [here](https://github.com/apache/iceberg-python/blob/a275ce5941dc85b1721562ea4d2e316e2b36b28e/pyiceberg/table/update/snapshot.py#L144) looks like manifests which don't match are just ignored.